src/Compilers/Core/Portable/SourceGeneration/GeneratorState.cs

```diff
@@ -2,31 +2,73 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
 using System.Collections.Immutable;
+using System.Diagnostics;
 
 #nullable enable
 namespace Microsoft.CodeAnalysis
 {
     internal readonly struct GeneratorState
     {
         public GeneratorState(GeneratorInfo info)
-            : this(info, ImmutableDictionary<GeneratedSourceText, SyntaxTree>.Empty)
+            : this(info, ImmutableArray<GeneratedSourceText>.Empty, ImmutableArray<SyntaxTree>.Empty, ImmutableArray<Diagnostic>.Empty, syntaxReceiver: null, exception: null)
         {
         }
 
-        private GeneratorState(GeneratorInfo info, ImmutableDictionary<GeneratedSourceText, SyntaxTree> sources)
+        private GeneratorState(GeneratorInfo info, ImmutableArray<GeneratedSourceText> sourceTexts, ImmutableArray<SyntaxTree> trees, ImmutableArray<Diagnostic> diagnostics, ISyntaxReceiver? syntaxReceiver, Exception? exception)
         {
-            this.Sources = sources;
+            this.SourceTexts = sourceTexts;
+            this.Trees = trees;
             this.Info = info;
+            this.Diagnostics = diagnostics;
+            this.SyntaxReceiver = syntaxReceiver;
+            this.Exception = exception;
         }
 
-        internal ImmutableDictionary<GeneratedSourceText, SyntaxTree> Sources { get; }
+        internal ImmutableArray<GeneratedSourceText> SourceTexts { get; }
+
+        internal ImmutableArray<SyntaxTree> Trees { get; }
 
         internal GeneratorInfo Info { get; }
 
-        internal GeneratorState WithSources(ImmutableDictionary<GeneratedSourceText, SyntaxTree> sources)
+        internal ISyntaxReceiver? SyntaxReceiver { get; }
+
+        internal Exception? Exception { get; }
+
+        internal ImmutableArray<Diagnostic> Diagnostics { get; }
+
+        internal GeneratorState SetReceiver(ISyntaxReceiver syntaxReceiver)
+        {
+            return new GeneratorState(this.Info,
+                                      sourceTexts: this.SourceTexts,
+                                      trees: this.Trees,
+                                      diagnostics: this.Diagnostics,
+                                      syntaxReceiver: syntaxReceiver,
+                                      exception: null);
+        }
+
+        internal GeneratorState SetResult(ImmutableArray<GeneratedSourceText> sourceTexts,
```

Comment:
Consider using `With` here as the method prefix. That is the more common pattern in the code base for immutable style set methods.